### PR TITLE
PLAT-26 Check Java Version is supported

### PR DESCRIPTION
**Pull request by cyrusv**
Warn users if we suspect they are using a Java version that is not
1.7 or 1.8, and fail fast if they are definitely using Java 9.

### CODE REVIEW

File: src/oss/confluent.sh [hunk 1]
```diff
@@ -154,6 +154,20 @@ requirements() {
...
+    local java_version=$(java -version 2>&1 | sed -n ';s/.* version "\(.*\)\.\(.*\)\..*"/\1\2/p;')
+    echo "Using Java Version $java_version"
+    # Warn users if it is not Java7 or Java8, or if we can't figure it out.
+    if [[ $java_version -ne 17 && $java_version -ne 18 ]]; then
+        echo "WARNING: Java version 1.8 is recommended:"
+        echo "  See http://confluent.io for detailed version and interoperability information"
+    fi
+    # Fail fast if the version is definitely Java9
+    if [[ $java_version = 19 ]]; then
+        echo "Java 9 is unsupported at this time: Confluent CLI will exit."
```
Comment (ybyzek) [inline src/oss/confluent.sh]: We need some better messaging here because AK supports Java 9.  Either allow users to run `confluent start kafka` even with java 9, or explain what the issue is (that the other components don't support it)

File: src/oss/confluent.sh [hunk 2]
```diff
@@ -154,6 +154,20 @@ requirements() {
     if [[ ${status} -ne 0 ]]; then
         FORMAT_CMD="xargs -0"
     fi
+
+    # Validate Java Version
+    local java_version=$(java -version 2>&1 | sed -n ';s/.* version "\(.*\)\.\(.*\)\..*"/\1\2/p;')
+    echo "Using Java Version $java_version"
+    # Warn users if it is not Java7 or Java8, or if we can't figure it out.
+    if [[ $java_version -ne 17 && $java_version -ne 18 ]]; then
+        echo "WARNING: Java version 1.8 is recommended:"
```
Comment (ybyzek) [inline src/oss/confluent.sh]: Inconsistency in formatting of java version: message here says 1.8, message below says 9.

File: src/oss/confluent.sh (hunk 1 above)
Comment (ybyzek) [inline src/oss/confluent.sh]: Also, add link to the version and interoperability information in docs: https://docs.confluent.io/current/installation/versions-interoperability.html

File: src/oss/confluent.sh [hunk 3]
```diff
@@ -154,6 +154,20 @@ requirements() {
...
+        echo "  See http://confluent.io for detailed version and interoperability information"
```
Comment (ybyzek) [inline src/oss/confluent.sh]: Add link: https://docs.confluent.io/current/installation/versions-interoperability.html

File: src/oss/confluent.sh [hunk 4]
```diff
@@ -911,7 +911,26 @@ list_command() {
...
+        return
+    fi
+    local java_version=$(java -version 2>&1 | sed -n ';s/.* version "\(.*\)\.\(.*\)\..*"/\1\2/p;')
+    # Warn users if it is not Java7 or Java8, or if we can't figure it out.
+    if [[ $java_version -ne 17 && $java_version -ne 18 ]]; then
+        echo "WARNING: Java version 1.8 is recommended:"
+        echo "  See https://docs.confluent.io/current/installation/versions-interoperability.html for detailed version and interoperability information"
+    fi
+    # Fail fast if the version is definitely Java9
+    if [[ $java_version = 19 ]]; then
```
Comment (wicknicks) [inline src/oss/confluent.sh]: java `10.0.1` was released last week. we might want to add that case here too.

File: src/oss/confluent.sh [hunk 5]
```diff
@@ -911,7 +911,26 @@ list_command() {
     fi
 }
 
+validate_java_version() {
+    local target_service=${1}
+    if [[ $target_service = "zookeeper" || target_service = "kafka" ]]; then
+        return
+    fi
+    local java_version=$(java -version 2>&1 | sed -n ';s/.* version "\(.*\)\.\(.*\)\..*"/\1\2/p;')
```
Comment (C0urante) [inline src/oss/confluent.sh]: Are we sure that the output of `java -version` is going to match this regex for different JVMs?

File: src/oss/confluent.sh (hunk 4 above)
Comment (wicknicks) [reply src/oss/confluent.sh]: Also, when running the sed command with Java 10 on my computer I get this:

```
$ /Library/Java/JavaVirtualMachines/jdk-10.0.1.jdk/Contents/Home/bin/java -version 2>&1 | sed -n ';s/.* version "\(.*\)\.\(.*\)\..*"/\1\2/p;'
100 2018-04-17
```

I think we should simplify the sed command. It will be easier to read/modify if the processing is done in multiple steps. A suggestion: 

```
java -version 2>&1  | head -n 1 | awk -F '"' '{print $2}' | awk -F'.' '{OFS = ""; print $1,$2}'
```

File: src/oss/confluent.sh (hunk 5 above)
Comment (cyrusv) [reply src/oss/confluent.sh]: Nope. Wasn't sure how I could test this, so the code is designed to "fail safe" and just show a warning if the regex parse fails. It will only hard-fail if it does parse correctly and finds an illegal version. If you have any ideas on how to test/explore this on more platforms, I'd definitely prefer to incorporate

File: src/oss/confluent.sh (hunk 4 above)
Comment (cyrusv) [reply src/oss/confluent.sh]: Will do

File: src/oss/confluent.sh (hunk 5 above)
Comment (C0urante) [reply src/oss/confluent.sh]: Nah, that sounds sufficient. Couldn't find any specs on what the output of `java -version` should be across JVMs after a (brief) search so this seems like the right call.

Comment (sansanichfb) [reply src/oss/confluent.sh]: If you don't want to rely on the os-specific output of `java -version` command - you can also create and execute jar file which runs `System.getProperty("java.version")` but it seems like overkill here.

Comment (cyrusv) [reply src/oss/confluent.sh]: Yes I think ideally we'd just put that in the `main` of each service to have a single point of truth, but this is meant to help our newbie 15-minute customers so this is sufficient.

File: src/oss/confluent.sh [hunk 6]
```diff
@@ -911,7 +911,26 @@ list_command() {
...
+    # Warn users if it is not Java7 or Java8, or if we can't figure it out.
+    if [[ $java_version -ne 17 && $java_version -ne 18 ]]; then
+        echo "WARNING: Java version 1.8 is recommended:"
+        echo "  See https://docs.confluent.io/current/installation/versions-interoperability.html for detailed version and interoperability information"
+    fi
+    # Fail fast if the version is definitely Java9
+    if [[ $java_version = 19 || $java_version = 110]]; then
+        echo "Java 1.9 is unsupported at this time: Confluent CLI will exit."
```
Comment (ybyzek) [inline src/oss/confluent.sh]: Before exit, can we add the same doc link as above?

Comment (cyrusv) [reply src/oss/confluent.sh]: It will execute both conditions, so it does the first block (prints link) and second block (exits)

Comment (ybyzek) [reply src/oss/confluent.sh]: Good point, I missed that.

File: src/oss/confluent.sh [hunk 7]
```diff
@@ -911,7 +911,26 @@ list_command() {
...
+        return
+    fi
+    local java_version=$(java -version 2>&1 |  head -n 1 | awk -F '"' '{print $2}' | awk -F'.' '{OFS = ""; print $1,$2}')
+    # Warn users if it is not Java7 or Java8, or if we can't figure it out.
+    if [[ $java_version -ne 17 && $java_version -ne 18 ]]; then
+        echo "WARNING: Java version 1.8 is recommended:"
+        echo "  See https://docs.confluent.io/current/installation/versions-interoperability.html for detailed version and interoperability information"
+    fi
+    # Fail fast if the version is definitely Java9
+    if [[ $java_version = 19 || $java_version = 110]]; then
```
Comment (wicknicks) [inline src/oss/confluent.sh]: I think Java 9 will return "90" with your command, and Java 10 will return "100". We can just do `$java_version -gt 18` if you are not sure of the output of the version command.

File: src/oss/confluent.sh [hunk 8]
```diff
@@ -911,7 +911,26 @@ list_command() {
...
 }
 
+validate_java_version() {
+    local target_service=${1}
+    if [[ $target_service = "zookeeper" || target_service = "kafka" ]]; then
+        return
+    fi
+    local java_version=$(java -version 2>&1 |  head -n 1 | awk -F '"' '{print $2}' | awk -F'.' '{OFS = ""; print $1,$2}')
+    # Warn users if it is not Java7 or Java8, or if we can't figure it out.
+    if [[ $java_version -ne 17 && $java_version -ne 18 ]]; then
```
Comment (kkonstantine) [inline src/oss/confluent.sh]: similar wrapping ${java_version}

File: src/oss/confluent.sh [hunk 9]
```diff
@@ -911,7 +911,26 @@ list_command() {
...
+  See https://docs.confluent.io/current/installation/versions-interoperability.html
+EOF
+    fi
+    if [[ $java_version -gt 18 ]]; then
+        die "Current Java version unsupported at this time: Confluent CLI will exit."
+    fi
+}
+
 start_command() {
+    validate_java_version ${1}
```
Comment (kkonstantine) [inline src/oss/confluent.sh]: same as below: `validate_java_version "${1}"`

File: src/oss/confluent.sh [hunk 10]
```diff
@@ -651,7 +651,35 @@ list_command() {
...
+    # 10.0.1 -> java version "10.0.1" 2018-04-17
+    # We need to match to the end of the line to prevent sed from printing the characters that do not match
+    local java_version=$(java -version 2>&1 | sed -E -n 's/.* version "([0-9]*).*$/\1/p')
+    # Warn users if it is not Java7 or Java8, or if we can't figure it out.
+    if [[ "${java_version}" -ge "9" ]]; then
+        cat <<EOF
+WARNING: Java version 1.8 is recommended:
+See https://docs.confluent.io/current/installation/versions-interoperability.html
+EOF
+        die "Current Java version unsupported at this time: Confluent CLI will exit."
```
Comment (kkonstantine) [inline src/oss/confluent.sh]: I think it's a good idea to print what's the current version that is not supported. 

Users might have multiple versions and could be not obvious which one the CLI is picking up. (Also reverse the order in which we print the error messages)

Suggestion: 
```java
   if [[ "${java_version}" -ge "9" ]]; then
        cat <<EOF
Current Java version '${java_version}' is unsupported at this time. Confluent CLI will exit.

EOF

        die "WARNING: Java version 1.8 is recommended. 
See https://docs.confluent.io/current/installation/versions-interoperability.html"
    fi
```

File: src/oss/confluent.sh [hunk 11]
```diff
@@ -651,7 +651,35 @@ list_command() {
...
+    fi
+}
+
 start_command() {
+    validate_java_version "${1}"
+    set_or_get_current
```
Comment (kkonstantine) [inline src/oss/confluent.sh]: This is not related to the current commit. Probably an artifact of rebasing

File: src/oss/confluent.sh [hunk 12]
```diff
@@ -651,7 +651,35 @@ list_command() {
...
+    echo "Using CONFLUENT_CURRENT: ${confluent_current}"
```
Comment (kkonstantine) [inline src/oss/confluent.sh]: This is not related to the current commit. Probably an artifact of rebasing
